### PR TITLE
[workloadmeta] Optimize `GetKubernetesPodForContainer` 

### PR DESCRIPTION
**Pull request by AliDatadog**
### What does this PR do?
This PR modifies the method `GetKubernetesPodForContainer` of the workloadmeta to leverage the field `Container.Owner.ID`. 

### Motivation
The `Owner` field can be leveraged to reduce the average complexity of the function from linear to constant time.

### Additional Notes
The owner field is not always accurate. 

### Possible Drawbacks / Trade-offs
I don't think there is any tradeoff because if a container is owned by a pod then both `pod.Containers` and `container.Owner` are set at the same time: https://github.com/DataDog/datadog-agent/blob/0682b43e9247a435ad62cd92c01aecb578f693d4/pkg/workloadmeta/collectors/internal/kubelet/kubelet.go#L115

### Describe how to test/QA your changes
If the tests pass then it should be ok.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/workloadmeta/testing/store.go [hunk 1]
```diff
@@ -75,23 +75,39 @@ func (s *Store) GetKubernetesPod(id string) (*workloadmeta.KubernetesPod, error)
 }
 
 // GetKubernetesPodForContainer returns a KubernetesPod that contains the
-// specified containerID.
+// specified containerID. The ownership information might lag because the kubelet check sets the owner field
+// but a container might be stored by CRI checks. Thus, the function might return an error when the pod
+// actually exists.
 func (s *Store) GetKubernetesPodForContainer(containerID string) (*workloadmeta.KubernetesPod, error) {
```
Comment (vboulineau) [inline pkg/workloadmeta/testing/store.go:81]: This is only the `testing` implementation, not the real one.

File: pkg/workloadmeta/testing/store.go [hunk 2]
```diff
@@ -75,23 +75,39 @@ func (s *Store) GetKubernetesPod(id string) (*workloadmeta.KubernetesPod, error)
...
+	if !ok {
+		return nil, errors.NewNotFound(containerID)
+	}
+
+	container := containerEntity.(*workloadmeta.Container)
+	if container.Owner == nil || container.Owner.Kind != workloadmeta.KindKubernetesPod {
+		return nil, errors.NewNotFound(containerID)
+	}
+
+	podEntities, ok := s.store[workloadmeta.KindKubernetesPod]
```
Comment (vboulineau) [inline pkg/workloadmeta/testing/store.go:100]: This can be replace by `s.GetKubernetesPod(container.Owner.ID)`

**Review by vboulineau — CHANGES_REQUESTED** [review]

File: pkg/workloadmeta/testing/store.go (hunk 2 above)
Comment (AliDatadog) [reply pkg/workloadmeta/testing/store.go:98]: It should be fixed by now

**Review by vboulineau — APPROVED** [review]

File: pkg/workloadmeta/store.go [hunk 3]
```diff
@@ -318,21 +318,35 @@ func (s *store) ListProcessesWithFilter(filter ProcessFilterFunc) []*Process {
...
+	}
+
+	container := containerEntity.cached.(*Container)
+	if container.Owner == nil || container.Owner.Kind != KindKubernetesPod {
+		return nil, errors.NewNotFound(containerID)
+	}
+
+	podEntities, ok := s.store[KindKubernetesPod]
+	if !ok {
+		return nil, errors.NewNotFound(container.Owner.ID)
```
Comment (remeh) [inline pkg/workloadmeta/store.go:341]: Nit: let's say there is a bug and the content of `container` isn't what's expected after the L329 read in the map, especially the `container.Owner.ID` field. This error then reports on an ID which isn't the same as the one passed to the method call.
I would thus prefer a:
```suggestion
		return nil, errors.NewNotFound(containerID)
```

That's purely debug/troubleshooting-wise, and just a nit. Discard if you feel like it.

File: pkg/workloadmeta/store.go [hunk 4]
```diff
@@ -318,21 +318,35 @@ func (s *store) ListProcessesWithFilter(filter ProcessFilterFunc) []*Process {
...
+	}
+
+	pod, ok := podEntities[container.Owner.ID]
+	if !ok {
+		return nil, errors.NewNotFound(container.Owner.ID)
```
Comment (remeh) [inline pkg/workloadmeta/store.go:346]: Ditto.

File: pkg/autodiscovery/listeners/container_test.go [hunk 5]
```diff
@@ -274,6 +278,10 @@ func TestCreateContainerService(t *testing.T) {
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
 			listener, wlm := newContainerListener(t)
+
+			if tt.container != nil {
+				listener.Store().(*workloadmetatesting.Store).Set(tt.container)
+			}
 			if tt.pod != nil {
 				listener.Store().(*workloadmetatesting.Store).Set(tt.pod)
 			}
```
Comment (remeh) [inline pkg/autodiscovery/listeners/container_test.go:287]: There is not any chance of a conflict between the `tt.container` or the `tt.pod` value in the store here?

File: pkg/workloadmeta/testing/store.go [hunk 6]
```diff
@@ -113,21 +113,35 @@ func (s *Store) GetKubernetesPod(id string) (*workloadmeta.KubernetesPod, error)
...
+	if !ok {
+		return nil, errors.NewNotFound(container.Owner.ID)
```
Comment (remeh) [inline pkg/workloadmeta/testing/store.go:136]: Ditto.

File: pkg/workloadmeta/testing/store.go [hunk 7]
```diff
@@ -113,21 +113,35 @@ func (s *Store) GetKubernetesPod(id string) (*workloadmeta.KubernetesPod, error)
...
+	}
+
+	pod, ok := podEntities[container.Owner.ID]
+	if !ok {
+		return nil, errors.NewNotFound(container.Owner.ID)
```
Comment (remeh) [inline pkg/workloadmeta/testing/store.go:141]: Ditto.

**Review by remeh — APPROVED** [review]
Left a question and a small remark, but 👍

File: pkg/workloadmeta/store.go (hunk 3 above)
Comment (AliDatadog) [reply pkg/workloadmeta/store.go:341]: Indeed, it is not the same but IIRC the pod ID has a different format from the container ID. It makes it easier to identify whether the pod or the container is not found.

File: pkg/autodiscovery/listeners/container_test.go (hunk 5 above)
Comment (AliDatadog) [reply pkg/autodiscovery/listeners/container_test.go:287]: Unless they have the same ID there shouldn't be any issue there. Those tests are edited manually and the unit test will fail if the ID is incorrect.

**Pull request merged**